// hybrids/lib/template/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.html = html;
exports.svg = svg;

var _define = _interopRequireDefault(require("../define"));

var _core = require("./core");

var _resolve = _interopRequireDefault(require("./resolve"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function defineElements(elements) {
  (0, _define.default)(elements);
  return this;
}

function key(id) {
  this.id = id;
  return this;
}

var updates = new Map();

function create(parts, args, isSVG) {
  var update = function update(host) {
    var target = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : host;
    var id = (0, _core.createId)(parts, isSVG);
    var render = updates.get(id);

    if (!render) {
      render = (0, _core.compile)(parts, isSVG);
      updates.set(id, render);
    }

    render(host, target, args);
  };

  return Object.assign(update, {
    define: defineElements,
    key: key
  });
}

function html(parts) {
  for (var _len = arguments.length, args = new Array(_len > 1 ? _len - 1 : 0), _key = 1; _key < _len; _key++) {
    args[_key - 1] = arguments[_key];
  }

  return create(parts, args);
}

function svg(parts) {
  for (var _len2 = arguments.length, args = new Array(_len2 > 1 ? _len2 - 1 : 0), _key2 = 1; _key2 < _len2; _key2++) {
    args[_key2 - 1] = arguments[_key2];
  }

  return create(parts, args, true);
}

Object.assign(html, {
  resolve: _resolve.default
});
Object.assign(svg, {
  resolve: _resolve.default
});
