// hybrids/lib/cache.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getEntry = getEntry;
exports.get = get;
exports.set = set;
exports.invalidate = invalidate;

var _utils = require("./utils");

var entries = new WeakMap();

function getEntry(target, key) {
  var targetMap = entries.get(target);

  if (!targetMap) {
    targetMap = new Map();
    entries.set(target, targetMap);
  }

  var entry = targetMap.get(key);

  if (!entry) {
    entry = {
      target: target,
      key: key,
      value: undefined,
      deps: new Set(),
      state: 1,
      checksum: 0
    };
    targetMap.set(key, entry);
  }

  return entry;
}

function calculateChecksum(_ref) {
  var state = _ref.state,
      deps = _ref.deps;
  var checksum = state;
  deps.forEach(function (entry) {
    // eslint-disable-next-line no-unused-expressions
    entry.target[entry.key];
    checksum += entry.state;
  });
  return checksum;
}

var context = null;

function get(target, key, getter) {
  var entry = getEntry(target, key);

  if (context === entry) {
    context = null;
    throw Error("[cache] Circular '".concat(key, "' get invocation in '").concat((0, _utils.stringifyElement)(target), "'"));
  }

  if (context) {
    context.deps.add(entry);
  }

  var parentContext = context;
  context = entry;

  if (entry.checksum && entry.checksum === calculateChecksum(entry)) {
    context = parentContext;
    return entry.value;
  }

  entry.deps.clear();

  try {
    var nextValue = getter(target, entry.value);

    if (nextValue !== entry.value) {
      entry.state += 1;
      entry.value = nextValue;
    }

    entry.checksum = calculateChecksum(entry);
    context = parentContext;
  } catch (e) {
    context = null;
    throw e;
  }

  return entry.value;
}

function set(target, key, setter, value, callback) {
  if (context) {
    context = null;
    throw Error("[cache] Try to set '".concat(key, "' of '").concat((0, _utils.stringifyElement)(target), "' in get call"));
  }

  var entry = getEntry(target, key);
  var newValue = setter(target, value, entry.value);

  if (newValue !== entry.value) {
    entry.state += 1;
    entry.value = newValue;
    callback();
  }
}

function invalidate(target, key, clearValue) {
  if (context) {
    context = null;
    throw Error("[cache] Try to invalidate '".concat(key, "' in '").concat((0, _utils.stringifyElement)(target), "' get call"));
  }

  var entry = getEntry(target, key);
  entry.checksum = 0;

  if (clearValue) {
    entry.value = undefined;
  }
}
