// hybrids/lib/children.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = children;

function walk(node, fn, options) {
  var items = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : [];
  Array.from(node.children).forEach(function (child) {
    var hybrids = child.constructor.hybrids;

    if (hybrids && fn(hybrids)) {
      items.push(child);

      if (options.deep && options.nested) {
        walk(child, fn, options, items);
      }
    } else if (options.deep) {
      walk(child, fn, options, items);
    }
  });
  return items;
}

function children(hybridsOrFn) {
  var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {
    deep: false,
    nested: false
  };
  var fn = typeof hybridsOrFn === 'function' ? hybridsOrFn : function (hybrids) {
    return hybrids === hybridsOrFn;
  };
  return {
    get: function get(host) {
      return walk(host, fn, options);
    },
    connect: function connect(host, key, invalidate) {
      var observer = new MutationObserver(invalidate);
      var set = new Set();

      var childEventListener = function childEventListener(_ref) {
        var target = _ref.target;

        if (!set.size) {
          Promise.resolve().then(function () {
            var list = host[key];

            for (var i = 0; i < list.length; i += 1) {
              if (set.has(list[i])) {
                invalidate(false);
                break;
              }
            }

            set.clear();
          });
        }

        set.add(target);
      };

      observer.observe(host, {
        childList: true,
        subtree: !!options.deep
      });
      host.addEventListener('@invalidate', childEventListener);
      return function () {
        observer.disconnect();
        host.removeEventListener('@invalidate', childEventListener);
      };
    }
  };
}
