// hybrids/lib/render.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.update = update;
exports.default = render;

var _utils = require("./utils");

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function _typeof(obj) { if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

var map = new WeakMap();
var cache = new WeakMap();
var FPS_THRESHOLD = 1000 / 60; // 60 FPS ~ 16,67ms time window

var queue = [];

function update() {
  var index = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
  var startTime = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 0;

  if (startTime && performance.now() - startTime > FPS_THRESHOLD) {
    requestAnimationFrame(function () {
      return update(index);
    });
  } else {
    var target = queue[index];
    var nextTime = performance.now();

    if (!target) {
      (0, _utils.shadyCSS)(function (shady) {
        return queue.forEach(function (t) {
          return shady.styleSubtree(t);
        });
      });
      queue = [];
    } else {
      if (map.has(target)) {
        var key = map.get(target);
        var prevUpdate = cache.get(target);

        try {
          var nextUpdate = target[key];

          if (nextUpdate !== prevUpdate) {
            cache.set(target, nextUpdate);
            nextUpdate();
            if (!prevUpdate) (0, _utils.shadyCSS)(function (shady) {
              return shady.styleElement(target);
            });
          }
        } catch (e) {
          update(index + 1, nextTime);
          throw e;
        }
      }

      update(index + 1, nextTime);
    }
  }
}

function addToQueue(event) {
  var target = event.composedPath()[0];

  if (target === event.currentTarget) {
    if (!queue[0]) {
      requestAnimationFrame(function () {
        return update();
      });
    }

    if (queue.indexOf(target) === -1) {
      queue.push(target);
    }
  }
}

function render(_get) {
  var customOptions = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

  if (typeof _get !== 'function') {
    throw TypeError("[render] The first argument must be a function: ".concat(_typeof(_get)));
  }

  var options = _objectSpread({
    shadowRoot: true
  }, customOptions);

  return {
    get: function get(host) {
      var fn = _get(host);

      return function () {
        return fn(host, options.shadowRoot ? host.shadowRoot : host);
      };
    },
    connect: function connect(host, key) {
      if (map.has(host)) {
        throw Error("[render] Render factory already used in '".concat(map.get(host), "' key"));
      }

      if (options.shadowRoot && !host.shadowRoot) {
        var shadowRootInit = {
          mode: 'open'
        };

        if (_typeof(options.shadowRoot) === 'object') {
          Object.assign(shadowRootInit, options.shadowRoot);
        }

        host.attachShadow(shadowRootInit);
      }

      host.addEventListener('@invalidate', addToQueue);
      map.set(host, key);
      return function () {
        host.removeEventListener('@invalidate', addToQueue);
        map.delete(host);
      };
    }
  };
}
